--- o/init	2020-04-02 16:19:07.000000000 +0800
+++ i/init	2020-04-02 16:15:52.000000000 +0800
@@ -239,6 +239,44 @@
 
 export ANDROID_ROOT=/system
 
+MIR_ROOT=/realblk
+mkdir -p /android/${MIR_ROOT}/system
+mkdir -p /android/${MIR_ROOT}/data
+mount --bind /mnt/$SRC/data /android/${MIR_ROOT}/data
+mount --bind /android/system /android/${MIR_ROOT}/system
+
+cat >> /android/init.rc << EOF
+on fs
+    start trick
+
+service trick /sbin/trick.sh
+    oneshot
+    disabled
+EOF
+
+cat >> /android/sbin/trick.sh << EOF
+#!/system/bin/sh
+
+while true; do
+  if [ -d /sbin/.magisk ]; then
+    break
+  fi
+done
+
+mkdir -p /sbin/.magisk/block
+touch /sbin/.magisk/block/system
+touch /sbin/.magisk/block/data
+
+mkdir -p /sbin/.magisk/mirror/system
+mkdir -p /sbin/.magisk/mirror/data
+
+mount -o bind ${MIR_ROOT}/system /sbin/.magisk/mirror/system
+mount -o bind ${MIR_ROOT}/data /sbin/.magisk/mirror/data
+EOF
+
+chcon u:object_r:system_file:s0 /android/sbin/trick.sh
+chmod 755 /android/sbin/trick.sh
+
 exec ${SWITCH:-switch_root} /android /init
 
 # avoid kernel panic
